Single-tree-render codemod: transform ReactDom.render()

simison opened the pull request:
(Related to PR #19494 _"Single tree-rendering-issues"_)

- Adds transformation of `ReactDom.render()` to `context.primary/secondary`
- Ensures `context` and `next` params are defined for middlewares, but is careful not to add them twice
- Refactors codemod to by more DRY
   Since `transformReactDomRender()` and `transformRenderWithReduxStore()` would've been almost identical, I split that stuff up to three functions. There might be other callExpressions we might want to codemod in similar way so this should be easy to extend.
- Adds comments for clarity

### Test
```bash
npm run codemod single-tree-rendering \
./client/auth/controller.js \
./client/devdocs/controller.js \
./client/components/drop-zone/test/index.jsx \
./client/devdocs/controller.js \
./client/layout/error.jsx \
./client/lib/abtest/test-helper/index.js \
./client/my-sites/customize/controller.js \
./client/post-editor/controller.js
```

### Problems/questions
- [ ] `next()` are all over the place, e.g.:
    ```js
    oauthLogin: function() {
      if ( config.isEnabled( 'oauth' ) ) {
        if ( OAuthToken.getToken() ) {
          page( '/' );
        } else {
          ReactDom.render( <OAuthLogin />, document.getElementById( 'primary' ) );
        }
      } else {
        page( '/' );
      }
    },
    ```
    
    Turns into:
    ```js
    oauthLogin: function(context, next) {
      if ( config.isEnabled( 'oauth' ) ) {
        if ( OAuthToken.getToken() ) {
          page( '/' );
        } else {
          context.primary = <OAuthLogin />;
        }
      } else {
        page( '/' );
      }
      next();
    },
    ```
    → See https://github.com/Automattic/wp-calypso/pull/19567#issuecomment-343595307

- [ ] Better way to check if these functions are really called as middleware? I find checking for number of parameters kinda smart trick but maybe there's a better way?
    → Probably not worth the effort

simison commented:
@ockham here's something for early preview so that you can see what I'm working on `next()`. ;-)

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -15,8 +31,25 @@ export default function transformer( file, api ) {
+	 */
+	function ensureContextMiddleware( path ) {
+		// `context` param is already in
+		if ( path.value.params.indexOf( 'context' ) !== -1 ) {
```
There might be occurrences where it's called `ctx` :slightly_smiling_face:

ockham replied:
Aside, are there really instances of middleware that don't even have a `context` arg?

ockham replied:
Hmm, fair enough, there's an example in the PR desc.

simison replied:
Searching for missing `context`s was a bit tricky so I simply decided to add this to see if it adds it to anywhere else but this example. :-)

Seems like `ctx` [has been cleaned out already](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=ctx&type=). Good you remembered something like that. :-)

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -15,8 +31,25 @@ export default function transformer( file, api ) {
+	 */
 	function addNextMiddleware( path ) {
-		if ( path.value.params.length !== 1 ) {
+		if ( path.value.params.length > 1 ) {
```
see https://github.com/Automattic/wp-calypso/pull/19567/files#r149681361

simison replied:
Yep so `ensureContextMiddleware()` bit is run first and therefore when `addNextMiddleware()` runs, there's exactly one param (`context`).

ockham commented on bin/codemods/src/single-tree-rendering.js:
````diff
@@ -40,16 +73,79 @@ export default function transformer( file, api ) {
+	 * ```
+	 * context.primary = <Example />;
+	 * ```
+	 */
````
👍

ockham commented:
> Better way to check if these functions are really called as middleware? I find checking for number of parameters kinda smart trick but maybe there's a better way?

Good question -- I don't know the answer :slightly_smiling_face: You can try to find an alternative, but best to timebox it; might be not worth wasting a lot of time on. Use your best judgement whether this can be reasonably easily automated, or if it'll require some inevitable manual intervention 😄

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -35,14 +43,30 @@ export default function transformer( file, api ) {
+		return (
+			_.findIndex( parameters || [], parameter => {
+				( parameter.name || parameter ) === 'context';
+			} ) !== -1
```
`_.findIndex() !== -1` should be synonymous to [`_.some()`](https://lodash.com/docs/4.17.4#some) which is a bit more semantic :slightly_smiling_face:

ockham commented on bin/codemods/src/single-tree-rendering.js:
```diff
@@ -35,14 +43,30 @@ export default function transformer( file, api ) {
+	 */
+	function hasParam( parameters, parameter ) {
+		return (
+			_.findIndex( parameters || [], parameter => {
```
I'd make the `[]` the default value for `hasParam`'s `parameters` arg.

simison replied:
Right on, I'm clearly working too much with es5 files at work... 🙃

simison commented:
@ockham What are your thoughts on this:
```js
var wrapper = document.createElement( ‘div’ );
document.body.appendChild( wrapper );

ReactDom.render(
    React.createElement( AcceptDialog, { ... } ),
    wrapper
);
```
— at [client/lib/accept/index.js](https://github.com/Automattic/wp-calypso/blob/master/client/lib/accept/index.js#L16-L17).

To me it looks like we could let that fly outside the tree, but I'm lacking experience with React to tell surely. Alternatively it seems like we'd have to bake it into each views just like any other element. WDYT?

simison commented:
Another interesting component: [root-child](https://github.com/Automattic/wp-calypso/tree/f1f5dd0ba751505c73bcf73a76f86f52cf651df6/client/components/root-child).

Needs a closer look and testing [where it's used](https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=root-child&type=).

simison commented:
I modified this to cover bigger variety of middlewares where `ReactDom.render()` might appear and now it's catching most of them.

I also changed `addNextMiddleware()` to `ensureNextMiddleware()` so that we can safely run it multiple times for the same function block. Something we'll end up doing if `ReactDom.unmountComponentAtNode()` and `ReactDom.render()` are both inside the same function (See e.g. [post-editor/controller.js](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/controller.js#L58-L71)).

Situations where `next()` ends up in a wrong place are tricky:
```js
oauthLogin: function(context, next) {
  if ( config.isEnabled( 'oauth' ) ) {
    if ( OAuthToken.getToken() ) {
      page( '/' );
    } else {
      context.primary = <OAuthLogin />;
      // ← `next()` Should be here...
    }
  } else {
    page( '/' );
  }
  // ...but it's here ↓
  next();
}
```

[This](https://github.com/Automattic/wp-calypso/blob/master/client/auth/controller.js#L27-L37) one is the only place where I've noticed this to happen, but that doesn't mean it's not happening.

Just looking at diffs it's easy to ignore those and these can easily cause nasty bugs. I'll skip fixing this for this PR and let's keep observing if this happens more. The above one we can fix manually.

simison commented:
Aight, tested this a bit more. Ready on my behalf. :-)

You can test this like so:
```
npm run codemod single-tree-rendering \
./client/auth/controller.js \
./client/devdocs/controller.js \
./client/components/drop-zone/test/index.jsx \
./client/devdocs/controller.js \
./client/layout/error.jsx \
./client/lib/abtest/test-helper/index.js \
./client/my-sites/customize/controller.js \
./client/post-editor/controller.js
```

You'll see a couple errors which are expected, those file we'll have to figure out manually.

ockham commented:
>You'll see a couple errors which are expected, those file we'll have to figure out manually.

Since you've already found a bunch of those troublemakers, I think it'd be good to keep a list somewhere -- or maybe even a PR holding commits with fixes.

ockham commented:
Looking at the test diff prompts two more wishlist items for future PRs :slightly_smiling_face: 

```diff
diff --git a/client/post-editor/controller.js b/client/post-editor/controller.js
index 223f8c0..ccf5a9e 100644
--- a/client/post-editor/controller.js
+++ b/client/post-editor/controller.js
@@ -55,19 +55,17 @@ function determinePostType( context ) {
        return context.params.type;
 }
 
-function renderEditor( context ) {
-       ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
-       ReactDom.render(
-               React.createElement(
-                       ReduxProvider,
-                       { store: context.store },
-                       React.createElement( PostEditor, {
-                               user: user,
-                               userUtils: userUtils,
-                       } )
-               ),
-               document.getElementById( 'primary' )
+function renderEditor(context, next) {
+    ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );
+       context.primary = React.createElement(
+               ReduxProvider,
+               { store: context.store },
+               React.createElement( PostEditor, {
+                       user: user,
+                       userUtils: userUtils,
+               } )
        );
+       next();
 }
```

1. We want to get rid of `ReduxProvider` here. This is MTR legacy; if we only have a single tree, one `ReduxProvider` wrapping the root component (probably `Layout`) is enough.
2. We also want to get rid of `ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );`. Right now, I think our `makeLayout` helper is requiring `context.secondary` to be `null`, but I don't think we even require that if we go fully STR. If we don't need anything in `secondary`, we just don't set it at all.

simison commented:
>We want to get rid of ReduxProvider here. 

:thumbsup: Cool, this was new info to me.

> We also want to get rid of `ReactDom.unmountComponentAtNode( document.getElementById( 'secondary' ) );`

:thumbsup: Yep, I started writing this already, PRing it next up.
